### PR TITLE
feat(aws-amplify-react-native): withOAuth Loading

### PR DESCRIPTION
**Pull request by jessedoyle**
*Issue #, if available:*

https://github.com/aws-amplify/amplify-js/issues/3021

*Description of changes:*

* Add a `loading` prop to the `withOAuth` higher order component
  that defaults to `true`.
* Once the amplify session has been created, the `loading` prop
  is set to `false`.
* This allows wrapped components to make decisions and display
  animations to users when the HOC is loading.
* Resolves https://github.com/aws-amplify/amplify-js/issues/3021

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jessedoyle** [comment]
It looks like https://github.com/aws-amplify/amplify-js/pull/3005 was just merged and changed OAuth behaviour. I'll see if I can update this PR shortly!

**Comment by jessedoyle** [comment]
Okay - I think I've fixed up this PR. I'm a little uncertain about how to handle the loading state change when initiating logout (or if this is necessary).

**Comment by jessedoyle** [comment]
@manueliglesias - I've rebased master into this PR. Any news on the review? Thanks!

File: packages/aws-amplify-react-native/src/Auth/withOAuth.js [hunk 1]
```diff
@@ -40,6 +40,7 @@ export default (Comp) => {
             this.state = {
                 user: null,
                 error: null,
+                loading: true,
```
Comment (jordanranz) [inline packages/aws-amplify-react-native/src/Auth/withOAuth.js]: Can we set the initial state of loading to false, then move the loading state change into componentDidMount() so it encapsulates state change into where the async function is actually made?

**Comment by jordanranz** [comment]
Looks good to me besides one minor change. Thanks for contributing!

File: packages/aws-amplify-react-native/src/Auth/withOAuth.js (hunk 1 above)
Comment (jessedoyle) [reply packages/aws-amplify-react-native/src/Auth/withOAuth.js]: Yup - absolutely!

Comment (jessedoyle) [reply packages/aws-amplify-react-native/src/Auth/withOAuth.js]: @jordanranz - I'm just curious how you would prefer the handling of `setState`? 

For now I'm going to implement `setState` with a callback function in `componentDidMount`. But we could use a different pattern with `componentDidUpdate` instead.

**Comment by jessedoyle** [comment]
@jordanranz - I amended my last commit to include your requested change.

**Comment by jordanranz** [comment]
LGTM. Thanks for making that change.

File: packages/aws-amplify-react-native/src/Auth/withOAuth.js (hunk 1 above)
Comment (jordanranz) [reply packages/aws-amplify-react-native/src/Auth/withOAuth.js]: The callback function looks good! Thanks for making this change. Approving.

**Pull request merged**